internal/host/statemachine.go

```
@@ -26,33 +27,53 @@ func NewHostStateMachine(th *transitionHandler) stateswitch.StateMachine {
 		PostTransition:   th.PostRegisterHost,
 	})
 
+	// Register host
 	sm.AddTransition(stateswitch.TransitionRule{
 		TransitionType:   TransitionTypeRegisterHost,
 		SourceStates:     []stateswitch.State{HostStatusInstalling, HostStatusInstallingInProgress},
 		DestinationState: HostStatusError,
 		PostTransition:   th.PostRegisterDuringInstallation,
 	})
 
+	// Installation failure
 	sm.AddTransition(stateswitch.TransitionRule{
 		TransitionType:   TransitionTypeHostInstallationFailed,
 		SourceStates:     []stateswitch.State{HostStatusInstalling, HostStatusInstallingInProgress},
 		DestinationState: HostStatusError,
 		PostTransition:   th.PostHostInstallationFailed,
 	})
 
+	// Cancel installation
 	sm.AddTransition(stateswitch.TransitionRule{
 		TransitionType:   TransitionTypeCancelInstallation,
 		SourceStates:     []stateswitch.State{HostStatusInstalling, HostStatusInstallingInProgress, HostStatusError},
 		DestinationState: HostStatusError,
 		PostTransition:   th.PostCancelInstallation,
 	})
 
+	// Reset host
 	sm.AddTransition(stateswitch.TransitionRule{
 		TransitionType:   TransitionTypeResetHost,
 		SourceStates:     []stateswitch.State{HostStatusError},
 		DestinationState: HostStatusResetting,
 		PostTransition:   th.PostResetHost,
 	})
 
+	// Install host
+	sm.AddTransition(stateswitch.TransitionRule{
+		TransitionType:   TransitionTypeInstallHost,
+		Condition:        th.IsValidRoleForInstallation,
+		SourceStates:     []stateswitch.State{HostStatusKnown},
+		DestinationState: HostStatusInstalling,
+		PostTransition:   th.PostInstallHost,
+	})
+
+	// Install disabled host will not do anything
+	sm.AddTransition(stateswitch.TransitionRule{
+		TransitionType:   TransitionTypeInstallHost,
+		SourceStates:     []stateswitch.State{HostStatusDisabled},
+		DestinationState: HostStatusDisabled,
```

Why do we need this transition?

---

Disable is a special case, we don't want to ignore hosts in disabled state while trying to install. so we don't want to change the state and don't want to return an error. 
Any other suggestion to do it? (there is a comment above the transition)

---

Will there be an error if no transition rule is fired?

---

Yes